src/Microsoft.Health.Fhir.Shared.Api/Features/Formatters/FhirJsonOutputFormatter.cs

```diff
@@ -61,6 +75,55 @@ public override async Task WriteResponseBodyAsync(OutputFormatterWriteContext co
             context.HttpContext.AllowSynchronousIO();
 
             HttpResponse response = context.HttpContext.Response;
+
+            var elementsSearchParameter = context.HttpContext.GetElementsSearchParameter(_logger);
+
+            Resource resource = null;
+
+            if (context.Object is Hl7.Fhir.Model.Bundle)
+            {
+                var bundle = context.Object as Hl7.Fhir.Model.Bundle;
+                resource = bundle;
+
+                if (elementsSearchParameter != null && elementsSearchParameter.Any())
+                {
+                    // _elements is not supported for a raw resource, revert to using FhirJsonSerializer
+                    foreach (var rawBundleEntryComponent in bundle.Entry)
+                    {
+                        if (rawBundleEntryComponent is RawBundleEntryComponent)
+                        {
+                            rawBundleEntryComponent.Resource = ((RawBundleEntryComponent)rawBundleEntryComponent).ResourceElement.ToPoco<Resource>(_deserializer);
+                        }
+                    }
+                }
+                else if (bundle.Entry.All(x => x is RawBundleEntryComponent))
+                {
+                    await _bundleSerializer.Serialize(context.Object as Hl7.Fhir.Model.Bundle, context.HttpContext.Response.Body);
+                    return;
+                }
+            }
+            else if (context.Object is RawResourceElement)
+            {
+                if (elementsSearchParameter != null && elementsSearchParameter.Any())
+                {
+                    // _elements is not supported for a raw resource, revert to using FhirJsonSerializer
+                    resource = (context.Object as RawResourceElement).ToPoco<Resource>(_deserializer);
+                }
+                else
+                {
+                    using (TextWriter textWriter = context.WriterFactory(response.Body, selectedEncoding))
+                    {
+                        textWriter.Write((context.Object as RawResourceElement).ResourceData);
```

Comment:
It would be more efficient to pass in the stream as a parameter to SerializeToJsonString (which would have a different name). The result does not need to be buffered as a string.

Reply:
Should be resolved now, but it feels more brittle and makes me a bit nervous